File: src/runtime/javascript/engine/nodes/table_pandas.ts
import { PortableRandom, portableSampleCount } from "../random";
import { Table, compileQuery, isMissing, toNumber } from "../table";
import { asBool, optionalFloat, parameterList, parseColumns, renameColumns, requireTable, resolveColumn, resolveColumns, scalarValue } from "./support/common";
import { filterRange, groupAggregate, groupByAggregate } from "./support/table_ops";
import type { ExecutionContext, NodeOutput } from "./support/types";

export function executeTablePandasNode(nodeType: string, params: Record<string, unknown>, upstream: unknown, context: ExecutionContext): NodeOutput | null {
  const tableResult: Table | null = null;
  const plotResult = null;
  const exportResult: string | null = null;
  const table = (): Table => requireTable(upstream, "Table input");

  switch (nodeType) {
    case "table.concat": {
      const axis = Number(params.axis ?? 0);
      if (axis !== 0 && axis !== 1) throw new Error("Concat axis must be 0 or 1");
      const inputs = upstream as Record<string, unknown>;
      const left = requireTable(inputs.left, "Concat input left");
      const right = requireTable(inputs.right, "Concat input right");
      const value = left.concat(right, axis as 0 | 1, asBool(params.ignoreIndex ?? axis === 0));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.select_columns": {
      const value = table().selectColumns(parseColumns(params.columns, table().columns.length));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.column_math": {
      const frame = table();
      const columns = resolveColumns(frame, params.columns);
      if (!columns.length) throw new Error("Column math requires at least one target column");
      const operation = String(params.operation ?? "multiply");
      if (!["add", "subtract", "multiply", "divide", "power", "absolute", "negate"].includes(operation)) {
        throw new Error(`Unsupported column math operation: ${operation}`);
      }
      const operand = Number(params.operand ?? 1);
      if (!Number.isFinite(operand)) throw new Error("Column math operand must be finite");
      if (operation === "divide" && operand === 0) throw new Error("Column math cannot divide by zero");
      const selected = new Set(columns.map((column) => frame.columnIndex(column)));
      const value = new Table(frame.columns, frame.rows().map((row) => row.map((cell, columnIndex) => {
        if (!selected.has(columnIndex) || isMissing(cell)) return cell;
        const number = toNumber(cell);
        if (number === null) throw new Error(`Column math requires numeric values in ${frame.columns[columnIndex]}`);
        if (operation === "add") return number + operand;
        if (operation === "subtract") return number - operand;
        if (operation === "multiply") return number * operand;
        if (operation === "divide") return number / operand;
        if (operation === "power") return number ** operand;
        if (operation === "absolute") return Math.abs(number);
        return -number;
      })));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.absolute": {
      const value = table().abs();
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.transpose": {
      const value = table().transpose();
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.slice": {
      const frame = table();
      const part = (prefix: string): { start: number | null; stop: number | null; step: number } => {
        const start = params[`${prefix}Start`];
        const stop = params[`${prefix}Stop`];
        const step = Math.trunc(Number(params[`${prefix}Step`] ?? 1));
        if (step === 0) throw new Error("Slice step cannot be zero");
        return { start: start === null || start === undefined || start === "" ? null : Number(start), stop: stop === null || stop === undefined || stop === "" ? null : Number(stop), step };
      };
      const indexes = (length: number, slice: { start: number | null; stop: number | null; step: number }): number[] => {
        const lower = slice.step < 0 ? -1 : 0;
        const upper = slice.step < 0 ? length - 1 : length;
        const norm = (value: number | null, isStart: boolean) => {
          if (value === null) return isStart ? (slice.step < 0 ? upper : lower) : (slice.step < 0 ? lower : upper);
          let result = Math.trunc(value); if (result < 0) result += length;
          return Math.min(upper, Math.max(lower, result));
        };
        const start = norm(slice.start, true); const stop = norm(slice.stop, false); const result: number[] = [];
        if (slice.step > 0) for (let i = start; i < stop; i += slice.step) result.push(i);
        else for (let i = start; i > stop; i += slice.step) result.push(i);
        return result;
      };
      const row = part("row"); const column = part("column");
      const value = frame.selectColumns(indexes(frame.columnCount, column)).takeRows(indexes(frame.rowCount, row));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.reset_index": {
      const value = table().resetIndex(asBool(params.drop ?? true));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.periodic_window": {
      const frame = table();
      const groupSize = Number(params.groupSize ?? 75);
      const count = Number(params.count ?? 25);
      if (groupSize < 1 || count < 1) throw new Error("Periodic window sizes must be positive");
      const position = String(params.position ?? "start");
      const offset = position === "end" ? groupSize - count : position === "offset" ? Number(params.offset ?? 0) : 0;
      if (offset < 0 || offset + count > groupSize) throw new Error("Periodic window must stay inside each group");
      const indexes: number[] = [];
      for (let base = 0; base < frame.rowCount; base += groupSize) {
        for (let r = base + offset; r < Math.min(base + offset + count, frame.rowCount); r += 1) indexes.push(r);
      }
      const value = frame.takeRows(indexes);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.periodic_tail_mean": {
      const frame = table();
      const value = frame.periodicTailMean(Number(params.groupSize ?? 25), Number(params.tailRows ?? 10), String(params.partialGroup ?? "include") as "include" | "drop" | "error");
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.row_chunks_to_columns": {
      const frame = table();
      const chunks = Number(params.chunks ?? 2);
      if (!Number.isInteger(chunks) || chunks < 1) throw new Error("Row chunks to columns requires chunks >= 1");
      const rows = frame.rows();
      const base = Math.floor(rows.length / chunks);
      const remainder = rows.length % chunks;
      const parts: unknown[][][] = [];
      let cursor = 0;
      for (let index = 0; index < chunks; index += 1) {
        const size = base + (index < remainder ? 1 : 0);
        parts.push(rows.slice(cursor, cursor + size));
        cursor += size;
      }
      const height = Math.max(0, ...parts.map((part) => part.length));
      const outputRows = Array.from({ length: height }, (_, rowIndex) => parts.flatMap((part) => part[rowIndex] ?? frame.columns.map(() => null)));
      const outputColumns = Array.from({ length: chunks }, (_, chunkIndex) => frame.columns.map((column) => `${column}_${chunkIndex + 1}`)).flat();
      const value = new Table(outputColumns, outputRows);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "stats.column_group_cv": {
      const frame = table();
      const groupSize = Number(params.groupSize ?? 50);
      if (!Number.isInteger(groupSize) || groupSize < 1) throw new Error("Column group CV requires groupSize >= 1");
      const rows = frame.rows();
      const groups: Array<Array<number | null>> = [];
      for (let start = 0; start < frame.columns.length; start += groupSize) {
        const stop = Math.min(start + groupSize, frame.columns.length);
        groups.push(rows.map((row) => {
          const values = row.slice(start, stop).map((item) => Number(item)).filter((item) => Number.isFinite(item));
          if (!values.length) return null;
          const mean = values.reduce((sum, item) => sum + item, 0) / values.length;
          if (mean === 0) return null;
          const variance = values.reduce((sum, item) => sum + (item - mean) ** 2, 0) / values.length;
          return Math.sqrt(variance) / mean;
        }));
      }
      return { outputs: { output: groups }, tableResult, plotResult, exportResult };
    }
    case "table.sort_index": {
      const axis = Number(params.axis ?? 0);
      const value = table().sortIndex(asBool(params.ascending ?? true), axis);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.difference": {
      const value = table().diff(Number(params.periods ?? 1), Number(params.axis ?? 0));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.filter_range": {
      const value = filterRange(table(), params);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.rename_columns": {
      const value = renameColumns(table(), params.names);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "table.pivot": {
      const frame = table();
      const index = resolveColumns(frame, params.index);
      const columns = resolveColumns(frame, params.columns);
      const values = resolveColumns(frame, params.values);
      if (index.length !== 1 || columns.length !== 1 || values.length !== 1) throw new Error("Pivot requires one row key, column key, and value column");
      const aggregate = String(params.aggregate ?? "mean");
      if (!["mean", "first", "max", "min"].includes(aggregate)) throw new Error("Unsupported pivot aggregate");
      let value = frame.pivot(index[0], columns[0], values[0], aggregate as "mean" | "first" | "max" | "min");
      if (asBool(params.resetIndex ?? true)) value = value;
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.dropna": {
      const frame = table();
      const how = String(params.how ?? "any");
      if (!["any", "all"].includes(how)) throw new Error("Drop missing values supports only any or all");
      const subset = resolveColumns(frame, params.subset);
      const value = frame.dropna(how as "any" | "all", subset);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.fillna": {
      const frame = table();
      const method = String(params.method ?? "value");
      let value: Table;
      if (method === "forward") value = frame.fillna("forward");
      else if (method === "backward") value = frame.fillna("backward");
      else if (method === "value") value = frame.fillna("value", scalarValue(params.value ?? "0"));
      else throw new Error(`Unsupported fill method: ${method}`);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.sort_values": {
      const frame = table();
      const columns = resolveColumns(frame, params.columns);
      if (!columns.length) throw new Error("Sort values requires at least one column");
      const naPosition = String(params.naPosition ?? "last");
      if (!["first", "last"].includes(naPosition)) throw new Error("naPosition must be first or last");
      const value = frame.sortValues(columns, asBool(params.ascending ?? true), naPosition as "first" | "last");
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.head": {
      const value = table().head(Number(params.n ?? 5));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.tail": {
      const value = table().tail(Number(params.n ?? 5));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.drop_duplicates": {
      const frame = table();
      const subset = resolveColumns(frame, params.subset);
      const keepRaw = String(params.keep ?? "first");
      const keep = keepRaw === "false" ? false : keepRaw as "first" | "last";
      const value = frame.dropDuplicates(subset, keep);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.sample": {
      const frame = table();
      const fraction = optionalFloat(params.fraction);
      const n = portableSampleCount(frame.rowCount, fraction, Number(params.n ?? 5));
      const value = frame.sample(n, asBool(params.replace ?? false), Number(params.randomState ?? 0));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.round": {
      const value = table().round(Number(params.decimals ?? 2));
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.describe": {
      const frame = table();
      const percentiles = parameterList(params.percentiles).map((item) => Number(item));
      const includeText = String(params.include ?? "").trim();
      const excludeText = String(params.exclude ?? "").trim();
      const include = includeText === "all" ? "all" as const : includeText ? parameterList(includeText).map(String) : null;
      const exclude = excludeText ? parameterList(excludeText).map(String) : null;
      const value = frame.describe(percentiles, include, exclude);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "pandas.query": {
      const expression = String(params.expression ?? "").trim();
      if (!expression) throw new Error("Query expression is required");
      const value = table().query(expression);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    default:
      return null;
  }
}

File: python/pydroid_flow/engine_parts/nodes/table_pandas.py
from __future__ import annotations

from typing import Any

import numpy as np
import pandas as pd

from ..analysis_nodes import _filter_range, _group_aggregate
from ..random_portable import _portable_sample_count, _portable_sample_indexes
from ..values import _as_bool, _optional_float, _parameter_list, _parse_columns, _rename_columns, _require_table, _resolve_columns, _scalar_value

NODE_TYPES = {
    "table.concat",
    "table.select_columns",
    "table.absolute",
    "table.transpose",
    "table.slice",
    "table.reset_index",
    "table.periodic_window",
    "table.periodic_tail_mean",
    "table.row_chunks_to_columns",
    "stats.column_group_cv",
    "table.sort_index",
    "table.difference",
    "table.filter_range",
    "table.rename_columns",
    "table.pivot",
    "table.group_aggregate",
    "table.groupby_aggregate",
    "pandas.dropna",
    "pandas.fillna",
    "pandas.sort_values",
    "pandas.head",
    "pandas.tail",
    "pandas.drop_duplicates",
    "pandas.sample",
    "pandas.round",
    "pandas.describe",
    "pandas.query",
}


def execute(
    node_type: str,
    params: dict[str, Any],
    upstream: Any,
    csv_text: str,
    input_files: list[dict[str, Any]],
    variables: dict[str, Any] | None = None,
) -> tuple[dict[str, Any], pd.DataFrame | None, str | None, str | None]:
    del csv_text, input_files, variables

    if node_type == "table.concat":
        axis = int(params.get("axis", 0))
        if axis not in {0, 1}:
            raise ValueError("Concat axis must be 0 or 1")
        value = pd.concat(
            [upstream["left"], upstream["right"]],
            axis=axis,
            ignore_index=_as_bool(params.get("ignoreIndex", axis == 0)),
        )
    elif node_type == "table.select_columns":
        table = _require_table(upstream, "Select columns")
        value = table.iloc[:, _parse_columns(params.get("columns"), len(table.columns))]
    elif node_type == "table.absolute":
        value = _require_table(upstream, "Absolute value").abs()
    elif node_type == "table.transpose":
        value = _require_table(upstream, "Transpose").transpose().reset_index(drop=True)
    elif node_type == "table.slice":
        table = _require_table(upstream, "Slice")

        def slice_part(prefix: str) -> slice:
            start = params.get(f"{prefix}Start")
            stop = params.get(f"{prefix}Stop")
            step = int(params.get(f"{prefix}Step", 1) or 1)
            if step == 0:
                raise ValueError("Slice step cannot be zero")
            return slice(None if start in {None, ""} else int(start), None if stop in {None, ""} else int(stop), step)

        value = table.iloc[slice_part("row"), slice_part("column")].copy()
    elif node_type == "table.reset_index":
        value = _require_table(upstream, "Reset index").reset_index(drop=_as_bool(params.get("drop", True)))
    elif node_type == "table.periodic_window":
        table = _require_table(upstream, "Periodic window")
        group_size = int(params.get("groupSize", 75))
        count = int(params.get("count", 25))
        if group_size < 1 or count < 1:
            raise ValueError("Periodic window sizes must be positive")
        position = str(params.get("position", "start"))
        offset = group_size - count if position == "end" else int(params.get("offset", 0)) if position == "offset" else 0
        if offset < 0 or offset + count > group_size:
            raise ValueError("Periodic window must stay inside each group")
        rows = [row for base in range(0, len(table), group_size) for row in range(base + offset, min(base + offset + count, len(table)))]
        value = table.iloc[rows].reset_index(drop=True)
    elif node_type == "table.periodic_tail_mean":
        table = _require_table(upstream, "Periodic tail mean")
        group_size = int(params.get("groupSize", 25))
        tail_rows = int(params.get("tailRows", 10))
        partial_group = str(params.get("partialGroup", "include"))
        if group_size < 1 or tail_rows < 1:
            raise ValueError("Periodic mean sizes must be positive integers")
        if tail_rows > group_size:
            raise ValueError("Periodic tailRows cannot exceed groupSize")
        if partial_group not in {"include", "drop", "error"}:
            raise ValueError(f"Unsupported partialGroup: {partial_group}")
        chunks: list[pd.Series] = []
        for start in range(0, len(table), group_size):
            end = min(start + group_size, len(table))
            actual_size = end - start
            if actual_size < group_size:
                if partial_group == "drop":
                    break
                if partial_group == "error":
                    raise ValueError(f"Periodic tail mean found incomplete final group with {actual_size} of {group_size} rows")
            tail_start = max(start, end - tail_rows)
            chunks.append(table.iloc[tail_start:end].mean(numeric_only=True))
        value = pd.DataFrame(chunks).reindex(columns=table.select_dtypes(include="number").columns)
    elif node_type == "table.row_chunks_to_columns":
        table = _require_table(upstream, "Row chunks to columns")
        chunks = int(params.get("chunks", 2))
        if chunks < 1:
            raise ValueError("Row chunks to columns requires chunks >= 1")
        arrays = np.array_split(table.to_numpy(), chunks, axis=0)
        frames = [
            pd.DataFrame(array, columns=[f"{column}_{index + 1}" for column in table.columns])
            for index, array in enumerate(arrays)
        ]
        value = pd.concat(frames, axis=1) if frames else table.iloc[0:0].copy()
    elif node_type == "stats.column_group_cv":
        table = _require_table(upstream, "Column group CV")
        group_size = int(params.get("groupSize", 50))
        if group_size < 1:
            raise ValueError("Column group CV requires groupSize >= 1")
        value = []
        for start in range(0, len(table.columns), group_size):
            group = table.iloc[:, start:start + group_size]
            means = group.mean(axis=1)
            stds = group.std(axis=1, ddof=0)
            value.append(pd.Series(np.where(means != 0, stds / means, np.nan), index=table.index))
    elif node_type == "table.sort_index":
        value = _require_table(upstream, "Sort index").sort_index(axis=int(params.get("axis", 0)), ascending=_as_bool(params.get("ascending", True)))
    elif node_type == "table.difference":
        table = _require_table(upstream, "Difference")
        value = table.diff(periods=int(params.get("periods", 1)), axis=int(params.get("axis", 0)))
    elif node_type == "table.filter_range":
        value = _filter_range(_require_table(upstream, "Range filter"), params)
    elif node_type == "table.rename_columns":
        value = _rename_columns(_require_table(upstream, "Rename columns"), params.get("names"))
    elif node_type == "table.pivot":
        table = _require_table(upstream, "Pivot")
        index = _resolve_columns(table, params.get("index"))
        columns = _resolve_columns(table, params.get("columns"))
        values = _resolve_columns(table, params.get("values"))
        if len(index) != 1 or len(columns) != 1 or len(values) != 1:
            raise ValueError("Pivot requires one row key, column key, and value column")
        aggregate = str(params.get("aggregate", "mean"))
        if aggregate not in {"mean", "first", "max", "min"}:
            raise ValueError("Unsupported pivot aggregate")
        value = table.pivot_table(index=index[0], columns=columns[0], values=values[0], aggfunc=aggregate).sort_index().sort_index(axis=1)
        value.columns = [str(column) for column in value.columns]
        if _as_bool(params.get("resetIndex", True)):
            value = value.reset_index()
    elif node_type == "pandas.dropna":
        table = _require_table(upstream, "Drop missing values")
        how = str(params.get("how", "any"))
        if how not in {"any", "all"}:
            raise ValueError("Drop missing values supports only any or all")
        subset = _resolve_columns(table, params.get("subset")) or None
        value = table.dropna(how=how, subset=subset).reset_index(drop=True)
    elif node_type == "pandas.fillna":
        table = _require_table(upstream, "Fill missing values")
        method = str(params.get("method", "value"))
        if method == "forward":
            value = table.ffill()
        elif method == "backward":
            value = table.bfill()
        elif method == "value":
            value = table.fillna(_scalar_value(params.get("value", "0")))
        else:
            raise ValueError(f"Unsupported fill method: {method}")
    elif node_type == "pandas.sort_values":
        table = _require_table(upstream, "Sort values")
        columns = _resolve_columns(table, params.get("columns"))
        if not columns:
            raise ValueError("Sort values requires at least one column")
        na_position = str(params.get("naPosition", "last"))
        if na_position not in {"first", "last"}:
            raise ValueError("naPosition must be first or last")
        value = table.sort_values(by=columns, ascending=_as_bool(params.get("ascending", True)), na_position=na_position).reset_index(drop=True)
    elif node_type == "pandas.head":
        value = _require_table(upstream, "Head").head(int(params.get("n", 5))).reset_index(drop=True)
    elif node_type == "pandas.tail":
        value = _require_table(upstream, "Tail").tail(int(params.get("n", 5))).reset_index(drop=True)
    elif node_type == "pandas.drop_duplicates":
        table = _require_table(upstream, "Drop duplicates")
        subset = _resolve_columns(table, params.get("subset")) or None
        keep_raw = str(params.get("keep", "first"))
        keep: Any = False if keep_raw == "false" else keep_raw
        value = table.drop_duplicates(subset=subset, keep=keep, ignore_index=_as_bool(params.get("ignoreIndex", True)))
    elif node_type == "pandas.sample":
        table = _require_table(upstream, "Sample")
        fraction = _optional_float(params.get("fraction"))
        replace = _as_bool(params.get("replace", False))
        count = _portable_sample_count(len(table), fraction, params.get("n", 5))
        indexes = _portable_sample_indexes(len(table), count, replace, int(params.get("randomState", 0)))
        value = table.iloc[indexes].copy()
        if _as_bool(params.get("ignoreIndex", True)):
            value = value.reset_index(drop=True)
    elif node_type == "pandas.round":
        value = _require_table(upstream, "Round").round(decimals=int(params.get("decimals", 2)))
    elif node_type == "pandas.describe":
        table = _require_table(upstream, "Describe")
        percentiles = [float(item) for item in _parameter_list(params.get("percentiles"))]
        include_text = str(params.get("include", "") or "").strip()
        exclude_text = str(params.get("exclude", "") or "").strip()
        include: Any = "all" if include_text == "all" else _parameter_list(include_text) or None
        exclude: Any = _parameter_list(exclude_text) or None
        value = table.describe(percentiles=percentiles or None, include=include, exclude=exclude).reset_index().rename(columns={"index": "statistic"})
    elif node_type == "pandas.query":
        table = _require_table(upstream, "Query")
        expression = str(params.get("expression", "")).strip()
        if not expression:
            raise ValueError("Query expression is required")
        value = table.query(expression).reset_index(drop=True)
    elif node_type == "table.group_aggregate":
        value = _group_aggregate(_require_table(upstream, "Group aggregate"), params)
    elif node_type == "table.groupby_aggregate":
        table = _require_table(upstream, "Groupby aggregate")
        group_by = _resolve_columns(table, params.get("groupBy"))
        if not group_by:
            raise ValueError("Groupby aggregate requires at least one grouping column")
        method = str(params.get("method", "mean"))
        if method not in {"mean", "median", "sum", "min", "max", "std", "count"}:
            raise ValueError(f"Unsupported groupby method: {method}")
        grouped = table.groupby(group_by, sort=True)
        if method == "count":
            value = grouped.size().reset_index(name="count")
        else:
            value = getattr(grouped, method)(numeric_only=True).reset_index()
    else:
        raise ValueError(f"Unsupported table/pandas node type: {node_type}")

    return {"output": value}, value if isinstance(value, pd.DataFrame) else None, None, None
